mypy/stubutil.py

```diff
@@ -751,7 +761,13 @@ def is_private_name(self, name: str, fullname: str | None = None) -> bool:
             return False
         if name == "_":
             return False
-        return name.startswith("_") and (not name.endswith("__") or name in self.IGNORED_DUNDERS)
+        if not name.startswith("_"):
+            return False
+        if self._all_ and name in self._all_:
+            return False
+        if name.startswith("__") and name.endswith("__"):
+            return name in self.IGNORED_DUNDERS
+        return True
```

Comment:
I think I prefer stubtest's heuristic here, which IIRC is something like:

1. If the runtime has `__all__`:
  a. If a name is included in `__all__`, treat it as public
  b. Else, treat the name as private
2. If the runtime doesn't have `__all__`:
  a. If the name doesn't start with `_`, treat it as public
  b. Else, if the name is a dunder, treat it as public unless it's one of the ones in the "excluded dunders" list
  c. Else, treat it as private

The key difference is that (IIUC) the current algorithm you have here would treat `bar` as a public name in this situation, whereas stubtest would treat it as private:

```py
__all__ = ["foo"]

foo = 2
bar = 3
```

Reply:
If we changed this algorithm, though, it would have an impact on existing tests such as `testIncludeClassNotInAll_inspect`. Probably best to leave it to another PR!

Reply:
Possibly we should change this, but that feels out of scope for this PR; the logic is mostly as it was before, except that names in `__all__` are now not treated as private. Your suggested change might be right, but it's better handled as an independent change.